File: sdk/example/raw/painting_node.dart
// Copyright 2015 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import "dart:math";
import 'dart:sky';

PaintingNode paintingNode = null;
Picture draw(int a, int r, int g, int b) {
  Rect bounds = new Rect.fromLTRB(0.0, 0.0, view.width, view.height);
  Size size = new Size(view.width, view.height);

  PictureRecorder recorder = new PictureRecorder();
  Canvas canvas = new Canvas(recorder, bounds);
  double radius = size.shortestSide * 0.45;

  Paint paint = new Paint()..color = new Color.fromARGB(a, r, g, b);
  canvas.drawCircle(size.center(Point.origin), radius, paint);

  if (paintingNode == null) {
    paintingNode = new PaintingNode();
    Paint innerPaint = new Paint()..color = new Color.fromARGB(a, 255 - r, 255 - g, 255 - b);
    PictureRecorder innerRecorder = new PictureRecorder();
    Canvas innerCanvas = new Canvas(innerRecorder, bounds);
    innerCanvas.drawCircle(size.center(Point.origin), radius * 0.5, innerPaint);

    paintingNode.setBackingDrawable(innerRecorder.endRecordingAsDrawable());
  }
  canvas.drawPaintingNode(paintingNode);

  return recorder.endRecording();
}

bool handleEvent(Event event) {
  if (event.type == "pointerdown") {
    view.picture = draw(255, 0, 0, 255);
    view.scheduleFrame();
    return true;
  }

  if (event.type == "pointerup") {
    view.picture = draw(255, 0, 255, 0);
    view.scheduleFrame();
    return true;
  }

  return false;
}

void main() {
  view.picture = draw(255, 0, 255, 0);
  view.scheduleFrame();

  view.setEventCallback(handleEvent);
}

File: sky/engine/core/painting/PaintingNode.cpp
// Copyright 2015 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "sky/engine/core/painting/PaintingNode.h"
#include "sky/engine/core/painting/Picture.h"

namespace blink {

class PaintingNodeDrawable : public SkDrawable {
public:
    static PassRefPtr<PaintingNodeDrawable> create(PassRefPtr<SkDrawable> skDrawable = nullptr);

    ~PaintingNodeDrawable() override;

    SkRect onGetBounds() override;
    void onDraw(SkCanvas* canvas) override;
    SkPicture* onNewPictureSnapshot() override;
    void set_drawable(PassRefPtr<SkDrawable> drawable) { m_drawable = drawable; }

private:
    PaintingNodeDrawable();
    explicit PaintingNodeDrawable(PassRefPtr<SkDrawable> skDrawable);
    RefPtr<SkDrawable> m_drawable;
};

// static
PassRefPtr<PaintingNodeDrawable> PaintingNodeDrawable::create(PassRefPtr<SkDrawable> skDrawable)
{
    return adoptRef(new PaintingNodeDrawable(skDrawable));
}

PaintingNodeDrawable::~PaintingNodeDrawable() {}


PaintingNodeDrawable::PaintingNodeDrawable(PassRefPtr<SkDrawable> skDrawable)
    : m_drawable(skDrawable)
{
}

SkPicture* PaintingNodeDrawable::onNewPictureSnapshot()
{
    if (!m_drawable)
        return nullptr;
    return m_drawable->newPictureSnapshot();
}

SkRect PaintingNodeDrawable::onGetBounds()
{
    if (!m_drawable)
        return SkRect::MakeEmpty();
    return m_drawable->getBounds();
}

void PaintingNodeDrawable::onDraw(SkCanvas* canvas)
{
    if (!m_drawable)
        return;
    return m_drawable->draw(canvas);
}




PassRefPtr<PaintingNode> PaintingNode::create()
{
    return adoptRef(new PaintingNode());
}

PaintingNode::PaintingNode()
    : m_paintingNodeDrawable(PaintingNodeDrawable::create())
{
}

void PaintingNode::setBackingDrawable(PassRefPtr<Drawable> drawable)
{
    m_paintingNodeDrawable->set_drawable(drawable->toSkia());
}

SkDrawable* PaintingNode::toSkia()
{
    return m_paintingNodeDrawable.get();
}

PassRefPtr<Picture> PaintingNode::newPictureSnapshot()
{
    ASSERT(m_paintingNodeDrawable);
    return Picture::create(
        adoptRef(m_paintingNodeDrawable->newPictureSnapshot()));
}

PaintingNode::~PaintingNode()
{
}

} // namespace blink
